### DISCUSSION TITLE
Can't get the sample code (for privacy group and private smart contract) to work locally

### DISCUSSION THREAD

**Discussion by rampatra** [web3j/web3j#1401] [Q&A] [answered]
I have set up Besu and Orion locally by following the guides here: https://besu.hyperledger.org/en/stable/Tutorials/Private-Network/Create-IBFT-Network/ and https://docs.orion.consensys.net/en/stable/HowTo/Install-Binaries/. 

And, I am passing the following flags:
```
besu --data-path=data --genesis-file=../genesis.json --rpc-http-enabled --rpc-http-api=ETH,NET,IBFT,WEB3,EEA,PRIV --host-allowlist="*" --rpc-http-cors-origins="all" --min-gas-price=0 --privacy-enabled --privacy-url=http://127.0.0.1:8888 --privacy-public-key-file=Orion/nodeKey.pub --privacy-onchain-groups-enabled=true
```

I am trying to use the code stated on http://docs.web3j.io/latest/privacy/privacy_support_web3j/, however, I am getting the below error:
```
org.web3j.protocol.exceptions.ClientConnectionException: Invalid response received: 400; {
  "jsonrpc" : "2.0",
  "id" : 2,
  "error" : {
    "code" : -32602,
    "message" : "Invalid params"
  }
}
```

My code so far is only this:
```
final Credentials ALICE = Credentials.create("8f2a55949038a9610f50fb23b5883af3b4ecb3c3bb792cbcefbd1542c692be63");
final Credentials BOB = Credentials.create("c87509a1c067bbde78beb793e6fa76530b6382a4c0241e5e4a9ec0a0f44dc0d3");
final Credentials CHARLIE = Credentials.create("ae6ae8e5ccbfb04590405997ee2d52d2b330726137b875053c36d94e974d162f");

final Base64String ENCLAVE_KEY_ALICE = Base64String.wrap("ApXYaIWJ8Mfm18YPHNdWLfZfdttcxfXPGOmKyshsC3g=");
final Base64String ENCLAVE_KEY_BOB = Base64String.wrap("vyGLN7JpYuk04z/aWlHIdAHXcdBj0I1F/BVlEEElSiU=");
final Base64String ENCLAVE_KEY_CHARLIE = Base64String.wrap("k2zXEin4Ip/qBGlRkJejnGWdP9cjkK+DAvKNW31L2C8=");

Besu nodeAlice = Besu.build(new HttpService("http://localhost:8545"));
Besu nodeBob = Besu.build(new HttpService("http://localhost:8546"));
Besu nodeCharlie = Besu.build(new HttpService("http://localhost:8547"));

PollingPrivateTransactionReceiptProcessor processor = new PollingPrivateTransactionReceiptProcessor(nodeAlice, 1000, 15);

// Generate a new random Base64 string for the privacy group ID
Base64String privacyGroupId = Base64String.wrap(generateRandomBytes(32));

// Create the privacy group
final String txHash = nodeAlice.privOnChainCreatePrivacyGroup(privacyGroupId, ALICE, ENCLAVE_KEY_ALICE,
        Collections.singletonList(ENCLAVE_KEY_BOB)).send().getTransactionHash();

// Wait for the transaction to be mined & get the receipt
TransactionReceipt receipt = processor.waitForTransactionReceipt(txHash);

// Find the privacy group based on the two members
Optional<PrivacyGroup> group = nodeAlice.privOnChainFindPrivacyGroup(Arrays.asList(ENCLAVE_KEY_ALICE, ENCLAVE_KEY_BOB))
        .send()
        .getGroups()
        .stream()
        .filter(x -> x.getPrivacyGroupId().equals(privacyGroupId))
        .findFirst();
 ```

Any tips? 

Do I need to enable some flags on my Orion nodes or Besu nodes?

**Comment by lucassaldanha** [comment]
I have found the issue. There was an update in some ABIs of the contracts that manage the privacy groups and that resulted in a change in the function hash. Flexible privacy groups have been released as early access, so the APIs are subject to change and we didn't catch this inconsistency in web3j. I'll discuss with the Besu team a way to improve our process to prevent this from happening again.

https://github.com/hyperledger/besu/blame/82847583355603cded899886140837f77375a246/privacy-contracts/src/main/solidity/DefaultOnChainPrivacyGroupManagementContract.sol#L46

I'll create a PR to fix this behaviour.

**Reply by lucassaldanha** [reply]
Actually, I have just noticed that this has been fixed by https://github.com/web3j/web3j/commit/79ded0284d5258e67bfc58a50b728672ae375f74

So the problem is that the latest released version of Web3j doesn't have the fix.

As soon as a new Web3j version is released it should be good to go.

**Comment by havdhe150692** [comment]
Hi @lucassaldanha  I'm having the same problem using version 4.9.4 of web3j, web3j-besu and using quorum-dev-quickstart. Is there any fix?

**Comment by macfarla** [comment]
Hi @havdhe150692 my guess is that it's a result of this breaking change https://github.com/hyperledger/besu/pull/4282 and again requires an update to web3j. 

The changes to the solidity contract can be viewed here https://github.com/hyperledger/besu/pull/4282/files#diff-0f68c916a9c9829baed118e7f2258dcf3644aaef3195fdfd12bd100394d19596

Looks like the change in web3j would be just to change bytes32 to bytes

**Comment by macfarla** [comment]
created an issue https://github.com/web3j/web3j/issues/1821

**Comment by NickSneo** [comment, marked as answer]
With latest web3j release private transactions are fixed, even we have integrations tests for private txs running using besu-dev-quickstart. So you can check example here - https://github.com/web3j/web3j/blob/master/integration-tests/src/test/java/org/web3j/protocol/besu/BesuPrivacyQuickstartIntegrationTest.java
